api/src/main/java/io/knotx/fragments/api/FragmentResult.java

```diff
@@ -34,32 +32,66 @@
   private static final String FRAGMENT_KEY = "fragment";
   private static final String TRANSITION_KEY = "transition";
   private static final String LOG_KEY = "log";
+  private static final String FAILURE_KEY = "error";
 
   private final Fragment fragment;
   private final String transition;
   private final JsonObject log;
+  private final FragmentOperationFailure error;
+
+  public static FragmentResult success(Fragment fragment, JsonObject log) {
+    return new FragmentResult(fragment, SUCCESS_TRANSITION, log);
+  }
+
+  public static FragmentResult success(Fragment fragment, String transition, JsonObject log) {
+    return new FragmentResult(fragment, transition, log);
+  }
+
+  public static FragmentResult fail(Fragment fragment, Throwable error) {
```

Comment:
I'm thinking where these factory methods should be used.
* Will this be called by an `Action` itself if it detects failure (like CB?) - then probably we could use some factory method to store logs as well?
* Or perhaps it should be called by an intermediate code (between action and task engine) which in case of exception tries to deliver a `FragmentResult` to TaskEngine anyway?

First case is in my opinion valid, second too - but we should probably consider when such fallback could occur, and think about use cases as well.

Reply:
Sure, I thought about those cases too. `Action` (`FragmentOperation`) can:
- end with `_success` transition which means SUCCESS
- end with `custom` transition which means SUCCESS
- end with `_error` transition which means that it failed but we handle an error in `Action`
- throws an exception

#### end with `_error` transition
In this case, we can:
- react to some condition and respond with an error, then we can return some error code and message
- handle some exception and respond with exception details

In both cases, we end with an error. 

However, the situation is a bit different for behaviours. Let's see the `Circuit Breaker Behaviour`:
- it calls a wrapped action, can be `HTTP action`
- the wrapped action responds with `_error` transition and `error` details (some error code)
- behaviour handles the `_error` transition and responds with `_fallback`, which means error too

So the `_fallback` transition means error too. However, it allows us to understand what is a root cause of the error:
- wrapped action
- behaviour itself

For behaviours I propose to follow such rules:
- it calls a wrapped action, can be `HTTP action`
- the wrapped action responds with `_error` transition and `error` details (some error code)
- behaviour handles the `_error` transition and responds with `_fallback` **and rewrites and error details into its error details**

#### throws an exception
Task Engine handles all exceptions and transforms them to `_error` transition with exception details.

Reply:
Thanks for a thorough explaination.

So if I understand it correctly, behaviour's result can have `errors` set and still respond with a `_success` transition? (Guess that will be the case for `InMemoryCache` action) 

In other words, does the `errors` property describe **problems that had already been handled** somehow? Can the receiver of `FragmentResult` **proceed with execution anyway**, as the resulting `fragment` and `transition` are set?

Reply:
Ok, I see what you mean.

In case of `doAction` error, a behaviour can:
- handle an error and respond with `_success` transition and `processing log`
- handle and error and respond with custom transition and `processing log` **or** `error log`
- handle and error and respond with `_error transition and `error log`
- rethrow an exception

> So if I understand it correctly, behaviour's result can have errors set and still respond with a _success transition?

I think it is **not** valid.

> In other words, does the errors property describe problems that had already been handled somehow? Can the receiver of FragmentResult proceed with execution anyway, as the resulting fragment and transition are set?

`Task Engine` calls a fragment operation once. There is no retry mechanizm or any timeout strategy - all those features are available in Actions. So when a fragment operation respons with an error, then `FragmentEvent` status is set to FAILURE. It is the `Task Engine` responsibility.
`Task Engine` is used by `FragmentsEngine` which **may** define (not now) some strategies for errors. In current approach, all details about errors will be serialised to `graph node processing log`.